File: limbo/library/src/main/java/co/windly/limbo/activity/base/LimboActivity.java
package co.windly.limbo.activity.base;

import android.os.Bundle;
import androidx.annotation.LayoutRes;
import androidx.annotation.Nullable;
import butterknife.ButterKnife;
import com.hannesdorfmann.mosby3.mvp.MvpActivity;

public abstract class LimboActivity<V extends LimboActivityView, P extends LimboActivityPresenter<V>> extends MvpActivity<V, P>
  implements LimboActivityView {

  //region Ui

  @LayoutRes
  protected abstract int getLayout();

  //endregion

  //region Lifecycle

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(getLayout());
    initializeViews();
    getPresenter().initializeManagers(this);
  }

  @Override
  protected void onDestroy() {
    getPresenter().clearDisposables();
    super.onDestroy();
  }

  //endregion

  //region View Injection

  protected void initializeViews() {
    ButterKnife.bind(this);
  }

  //endregion
}

File: limbo/library/src/main/java/co/windly/limbo/activity/base/LimboActivity.kt
package co.windly.limbo.activity.base

import android.os.Bundle
import androidx.annotation.LayoutRes
import co.windly.limbo.LimboPresenter
import co.windly.limbo.LimboView
import co.windly.limbo.disposable.LifecycleCompositeDisposable
import co.windly.limbo.disposable.lifecycleDestroyCompositeDisposable
import com.hannesdorfmann.mosby3.mvp.MvpActivity
import io.reactivex.disposables.Disposable

abstract class LimboActivity<V : LimboView, P : LimboPresenter<V>> : MvpActivity<V, P>(), LimboView {

  //region Reactive

  override val disposables: LifecycleCompositeDisposable
    by lifecycleDestroyCompositeDisposable()

  override fun addDisposable(disposable: Disposable): Boolean =
    disposables.add(disposable)

  override fun clearDisposables() {
    disposables.clear()
  }

  //endregion

  //region Ui

  @get:LayoutRes
  protected abstract val layout: Int

  //endregion

  //region Lifecycle

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    // Define content view.
    setContentView(layout)
  }

  override fun onDestroy() {

    // Clear presenter-bound disposables.
    getPresenter().clearDisposables()

    // Continue destroy'ing activity.
    super.onDestroy()
  }

  //endregion
}
